data/transactions/logic/opcodes.go

```diff
@@ -164,6 +164,8 @@ var OpSpecs = []OpSpec{
 
 	{0x70, "asset_holding_get", opAssetHoldingGet, assembleAssetHolding, disAssetHolding, twoInts, oneAny.plus(oneInt), 2, runModeApplication, opSize{1, 2, nil}},
 	{0x71, "asset_params_get", opAssetParamsGet, assembleAssetParams, disAssetParams, oneInt, oneAny.plus(oneInt), 2, runModeApplication, opSize{1, 2, nil}},
+
+	{0x72, "assert", opAssert, asmDefault, disDefault, oneInt, nil, 3, modeAny, opSizeDefault},
```

Comment:
I find this opcode to be redundant, as it's a synonymous to the byte code of `bnz [offset=100000]`, and I don't see massive usage of this command either. I would think that making it into a semantic sugar, or a macro would make more sense since the "saving" here are 2 bytes for a command that rarely used. Also keep in mind that since teal is a stack-language, the common pattern is to accumulate bunch of condition on the stack, and to make multiple && at the end. ( which are 1 byte long instruction, naturally )

If you do believe that this is an important command, I won't be blocking this as I don't see any real harm here other than "wasting" the instructions space.

Reply:
to me, it looks like the `assert` is a better fit for a functional programming rather then stack-based languages.

Reply:
The `bnz Infinity` option is an important point. It's possible that `assert` could be implemented as a psuedo-op in the assembler but we don't need an actual opcode for it.

Reply:
While `assert` is definitely not the most important opcode in the world (I only did it first because it was so simple, and want to check that I'm doing all the right things for addition) it has two important properties that I think make it a clear win.

1) It allows a TEAL programmer to express predicates *locally*.  They can avoid the && chain pattern, which is is error-prone.  You must skip the && on your first check, but nowhere else.  It runs all of your code when you could have exited early. You can't reliably take a predicate check clause and insert it somewhere else.  Once you've messed up the chain, even when the assembler catches your typing error (which it can only do if you've avoided jumps in your entire program), it is hard to figure out what the assembler is actually complaining about.

2) `assert` is just as compact as the && pattern (one byte per condition checked), so programmers have no reason to avoid it.  TEAL users do run into our resource limits.  If we given them a 3-byte alternative (jnz Infinity pseudo-op) they will avoid it, and continue to use the error-prone approach.

Reply:
I agree with both of your observations.